### ISSUE TITLE
Behaviour difference between app.map and app.mapWhen

### ISSUE THREAD

**Issue by chrisckc** [aspnet/AspNetCore#3762]
According to the docs:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-2.1#use-run-and-map

> When Map is used, the matched path segment(s) are removed from HttpRequest.Path and appended to HttpRequest.PathBase for each request.

The section about mapWhen does not say anything about what is does with HttpRequest.Path

**observation**
From my own testing i have found that 'mapWhen' does not do this, which is good because for my use case i need the HttpRequest.Path to be left untouched, however in production mode i need it to be removed as it is when using 'map'.

That little statement about map in the docs is easy to miss, a sensible assumption when discovering those methods would be that both methods behave the same way and don't modify the path.

To keep the code clean and more meaningful i would rather not wrap the 2 different methods inside an if statement for env.IsDevelopment() and factor out the common code etc.

**suggestion**
It would be great if both map and mapWhen could be extended with an option for removing the matched path segment eg. `bool removeMatchedPathSegment` keeping the defaults to match the current behaviour to avoid breaking changes to existing apps etc.

**workaround**
To achieve this in my app i will probably go down the route of creating my own configurable mapWhen middleware based on this:
https://github.com/aspnet/HttpAbstractions/blob/master/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/MapWhenMiddleware.cs
and then borrowing the code from MapMiddleware.cs to optionally remove the path segment.

I would rather not have to replicate framework code to achieve what i want to do so it would be great if this feature could be added for a future version. It would also make the behaviour more explicit to avoid any confusion.

**Comment by chrisckc** [comment]
Fair point, just add the 'removeMatchedPathSegment ' option to the map middleware defaulting to true.

In the meantime ill crate my own version of the map middleware rather then the mapWhen middleware.

**Comment by anurse** [comment]
We've moved this issue is in the Backlog milestone. This means that it is not going to happen for the coming release. We will re-assess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.
